### PR TITLE
allocator: make disk capacity threshold a setting

### LINKED ISSUE #97392: allocator: make disk capacity thresholds configurable
Currently the allocator will stop rebalancing towards a store when it is using >= 92.5% of disk.

https://github.com/cockroachdb/cockroach/blob/7e2df35a2f6bf7a859bb0539c8ca43c4e72ed260/pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go#L59-L59

The allocator will rebalance replicas away from at 95%.

https://github.com/cockroachdb/cockroach/blob/7e2df35a2f6bf7a859bb0539c8ca43c4e72ed260/pkg/kv/kvserver/allocator/base.go#L30-L30

These thresholds are hardcoded. This issue is to make these thresholds cluster settings so that they may be configured.

Jira issue: CRDB-24674

### PR DESCRIPTION
**Pull request by kvoli**
Previously, the store disk utilization was checked against a constant
threshold to determine if the store was a valid allocation target.

This commit adds two cluster settings to replace the constant.

`kv.allocator.max_disk_utilization_threshold`

Maximum disk utilization before a store will never be used as a
rebalance or allocation target and will actively have replicas moved off
of it.

`kv.allocator.rebalance_to_max_disk_utilization_threshold`

Maximum disk utilization before a store will never be used as a
rebalance target.

Resolves: https://github.com/cockroachdb/cockroach/issues/97392

Release note (ops change): Introduce two cluster settings to control disk
utilization thresholds for allocation.
`kv.allocator.rebalance_to_max_disk_utilization_threshold` Maximum disk
utilization before a store will never be used as a rebalance target.
`kv.allocator.max_disk_utilization_threshold` Maximum disk utilization
before a store will never be used as a rebalance or allocation target
and will actively have replicas moved off of it.

### CODE REVIEW

File: pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go [hunk 1]
```diff
@@ -167,6 +168,56 @@ var IOOverloadThresholdEnforcement = settings.RegisterEnumSetting(
...
+	"kv.allocator.max_disk_utilization_threshold",
+	"maximum disk utilization before a store will never be used as a rebalance "+
+		"or allocation target and will actively have replicas moved off of it",
+	defaultMaxDiskUtilizationThreshold,
+	func(f float64) error {
+		if f > 0.95 {
+			return errors.Errorf(
+				"Cannot set kv.allocator.max_disk_utilization_threshold greater than 0.95")
+		}
+		if f < 0.05 {
```
Comment (tbg) [inline pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go:186]: Interesting, we're allowing setting this to near zero? Almost seems as though we could abuse this to implement "gateway-only nodes" (nodes that don't hold any replicas) if we allowed setting this to 0.00001.

Just a drive-by.

Comment (kvoli) [reply pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go:186]: I didn't want to be too assuming w.r.t the threshold. The 0.95 max seems good but the 0.05 could be changed or removed.

**Review by andrewbaptist — CHANGES_REQUESTED** [review]
> Reviewed 3 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97409)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @tbg)

___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 1915 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NPsOMbnCJdZue9z1Cx5:-NPsOMbnCJdZue9z1Cx6:b-aiym10) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aea5175cf48c5e499bff1a7d2c1bee20bee0484/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L1915)):*
> ```Go
> func (a *Allocator) ScorerOptions(ctx context.Context) *RangeCountScorerOptions {
> 	return &RangeCountScorerOptions{
> 		DiskOptions:             a.DiskOptions(),
> ```

nit: I don't love the name `DiskOptions` as it isn't clear this is specifically about Capacity. `DiskCapacityOptions` would be better, but not necessarily worth changing everything for. There are other `Options` about disks that are not captured here (max throughput, number of ranges, ...). 
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go` line 186 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NOpZ5z0UNUu9FzOQ3k5-r1-186:-NPsMtcY0fixKud2i_3Q:bgxl0fn) ([raw file](https://github.com/cockroachdb/cockroach/blob/afff942d7125fe2ffa642ca068e184de80be255a/pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go#L186)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

I didn't want to be too assuming w.r.t the threshold. The 0.95 max seems good but the 0.05 could be changed or removed.
</blockquote></details>

This cap is too low. Setting it to 0.99 or 1.0 for safety so people don't think it's a percent and set it to something like '90' is useful, but this is too constraining. I could see a situation where the entire system is close to 95% full and the allocator is mostly disabled due to that. Normally, it doesn't make sense for the default value to be the same as the max allowed value. I also think it makes sense to add some warnings about setting this too low (but not add a check). If this is set at something like 0.5 because a user thinks they want to see their system balanced, it will not work as expected. The 0.5 will be hit, but the net effect is that other rebalancing will be broken and the system could even become unstable.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go` line 211 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NPsNjjdBccdWolRhhy-:-NPsNjjdBccdWolRhhy0:b-24dn56) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aea5175cf48c5e499bff1a7d2c1bee20bee0484/pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go#L211)):*
> ```Go
> 	},
> )
> ```

Similar to comment above, set the max allowed to be a higher value and add a note that this should be set lower than the previous setting and also not set too aggressively low. Also do we need a check that this is always strictly less than the one above? It might be OK to have there be no buffer, but as you note in the system this isn't ideal.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go` line 618 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NPsPaaaERlmXOfLEXZP:-NPsPaaaERlmXOfLEXZQ:bh1q1zx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aea5175cf48c5e499bff1a7d2c1bee20bee0484/pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go#L618)):*
> ```Go
> type DiskOptions struct {
> 	RebalanceToThreshold float64
> 	MaxThreshold         float64
> ```

nit: Consider renaming this `ShedThreshold` to clarify what happens when it hits this threshold.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go` line 1235 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NPsTv536Grqlyu9sseC:-NPsTv536Grqlyu9sseD:b6gsasp) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aea5175cf48c5e499bff1a7d2c1bee20bee0484/pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go#L1235)):*
> ```Go
> 			Capacity: roachpb.StoreCapacity{
> 				Capacity:   100,
> 				Available:  100 - int64(100*float64(defaultMaxDiskUtilizationThreshold)),
> ```

Consider writing this as: `100 - int64(100*float64(defaultRebalanceToMaxDiskUtilizationThreshold)) -1` as written it is "right on" the border, and could cause confusing off-by-one errors. 
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go` line 1247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NPsTS0-BqZtqNSSnzqS:-NPsTS0-BqZtqNSSnzqT:b-5dwy67) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aea5175cf48c5e499bff1a7d2c1bee20bee0484/pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go#L1247)):*
> ```Go
> 			Capacity: roachpb.StoreCapacity{
> 				Capacity:   100,
> 				Available:  (100 - int64(100*float64(defaultMaxDiskUtilizationThreshold))) / 2,
> ```

Consider writing as `100 - int64(100*float64(defaultMaxDiskUtilizationThreshold)) -1`
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go` line 1310 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NPsUc_Z0M5-41v0_7mV:-NPsUc__EeA4XTgJLm5J:b-ee2c1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aea5175cf48c5e499bff1a7d2c1bee20bee0484/pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go#L1310)):*
> ```Go
> 			a.Metrics,
> 		)
> 		if expResult := (i >= 3); expResult != result {
> ```

Why did the expected result change? Add a comment on why this is 3.

**Review by kvoli — COMMENTED** [review]
Thanks for taking a look. I've updated the patch in response to your feedback. It should be good for another round of review @andrewbaptist.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-:-NPxd8IB7LNcy770eu3e:b2oepde)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist, @kvoli, and @tbg)

___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 1915 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NPsOMbnCJdZue9z1Cx5:-NPwjkQj8Nx3lukdGzNk:b-hpq7hv) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aea5175cf48c5e499bff1a7d2c1bee20bee0484/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L1915)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

nit: I don't love the name `DiskOptions` as it isn't clear this is specifically about Capacity. `DiskCapacityOptions` would be better, but not necessarily worth changing everything for. There are other `Options` about disks that are not captured here (max throughput, number of ranges, ...). 
</blockquote></details>

Good point. Updated to `DiskCapacityOptions`
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go` line 186 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NOpZ5z0UNUu9FzOQ3k5-r1-186:-NPwkk0x7JnlRcJUjGlC:b-8ab2wp) ([raw file](https://github.com/cockroachdb/cockroach/blob/afff942d7125fe2ffa642ca068e184de80be255a/pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go#L186)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

This cap is too low. Setting it to 0.99 or 1.0 for safety so people don't think it's a percent and set it to something like '90' is useful, but this is too constraining. I could see a situation where the entire system is close to 95% full and the allocator is mostly disabled due to that. Normally, it doesn't make sense for the default value to be the same as the max allowed value. I also think it makes sense to add some warnings about setting this too low (but not add a check). If this is set at something like 0.5 because a user thinks they want to see their system balanced, it will not work as expected. The 0.5 will be hit, but the net effect is that other rebalancing will be broken and the system could even become unstable.
</blockquote></details>

Updated to >0.99. I also added a min check at 0 to prevent negative numbers. I don't think this actually matters but for completeness sake.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go` line 211 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NPsNjjdBccdWolRhhy-:-NPwl71I9v5SuoIHPjqu:b-wzghbt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aea5175cf48c5e499bff1a7d2c1bee20bee0484/pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go#L211)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Similar to comment above, set the max allowed to be a higher value and add a note that this should be set lower than the previous setting and also not set too aggressively low. Also do we need a check that this is always strictly less than the one above? It might be OK to have there be no buffer, but as you note in the system this isn't ideal.
</blockquote></details>

Updated to >0.99 check. Added note to not set higher than the other setting and vice versa on the other setting.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go` line 618 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NPsPaaaERlmXOfLEXZP:-NPxEZ9AE99iSNNbnAEN:bjkkdpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aea5175cf48c5e499bff1a7d2c1bee20bee0484/pkg/kv/kvserver/allocator/allocatorimpl/allocator_scorer.go#L618)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

nit: Consider renaming this `ShedThreshold` to clarify what happens when it hits this threshold.
</blockquote></details>

It is Shed + Block Allocation + Block Rebalancing. I'll update to be `ShedAndBlockAll`
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go` line 1235 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NPsTv536Grqlyu9sseC:-NPxbJ2sBr71eYSpvGF6:b-r88ne8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aea5175cf48c5e499bff1a7d2c1bee20bee0484/pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go#L1235)):*
> 100 - int64(100*float64(defaultMaxDiskUtilizationThreshold)) -1
I've updated this to be clearer.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go` line 1247 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NPsTS0-BqZtqNSSnzqS:-NPxbLGDA3aMDTHxq_cv:bs4q9op) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aea5175cf48c5e499bff1a7d2c1bee20bee0484/pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go#L1247)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Consider writing as `100 - int64(100*float64(defaultMaxDiskUtilizationThreshold)) -1`
</blockquote></details>

Ditto.
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go` line 1310 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-NPsUc_Z0M5-41v0_7mV:-NPxbLyo-0bAQ9aLUaU7:bpe02f7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2aea5175cf48c5e499bff1a7d2c1bee20bee0484/pkg/kv/kvserver/allocator/allocatorimpl/allocator_test.go#L1310)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Why did the expected result change? Add a comment on why this is 3.
</blockquote></details>

Previously the storepool wouldn't update the average Range Count for a StoreList if the candidate store didn't pass the maxCapacityCheck. I removed this to break the dependency on the constant. 

This logic was already redundant for rebalancing as we create equivalence classes based on equally valid/diverse and non-disk fullness. So the store list would never contain a full disk store to affect range count. For allocation, I updated the logic to be similar so we check earlier and remove candidates to then create a valid store list.

This test is was previously asserting that s3 which had 5 ranges (all stores sum = 27, num stores = 5) would be rebalanced away from because s4 and s5 would not have their range count added to the store list avg range count. The avg previously for the store list would be (7/5 = 1.4) so s3 we would expect shouldRebalanceAway to be true since it is greater than the mean. That is no longer the case so I've updated the test.

**Review by andrewbaptist — APPROVED** [review]
:lgtm: Thanks for the quick turnaround on this. We should absolutely merge this in because it will be helpful to have these knobs, but do you have an idea how a system would respond if it had an issue like the customer issue we had related to this? I assume it would keep the disks from getting full, but do you know if there would be other negative implications on rebalancing?

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-:-NPyATOFAqxsw5rIQe6O:b-oltfyk)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @kvoli and @tbg)

**Comment by kvoli** [comment]
> :lgtm: Thanks for the quick turnaround on this. We should absolutely merge this in because it will be helpful to have these knobs, but do you have an idea how a system would respond if it had an issue like the customer issue we had related to this? I assume it would keep the disks from getting full, but do you know if there would be other negative implications on rebalancing?
> 
> > Reviewed all commit messages.
> > *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97409#-:-NPyATOFAqxsw5rIQe6O:b-oltfyk)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @kvoli and @tbg)
> 
> 
> 
> 
> 

There would only be negative effects when there's a combination of other issues as well as full disk such as overload on every non full disk.

I think this should generally be pretty robust  unless misconfigured.

**Comment by kvoli** [comment]
bors r=andrewbaptist

**Comment by kvoli** [comment]
CI failure could be legit. I'll investigate tomorrow.

bors r-

**Comment by kvoli** [comment]
bors r=andrewbaptist
